test/cpp/qps/client_async.cc

```diff
@@ -949,10 +949,182 @@ std::unique_ptr<Client> CreateAsyncClient(const ClientConfig& config) {
+
+ protected:
+  size_t num_threads_;
+  std::mutex mu_;
+  std::condition_variable cv_;
+  std::vector<CallbackClientRpcData*> ctxs_;
+  size_t threads_done_;
+
+  void NotifyThread(size_t thread_idx) {
+    std::mutex* thread_mu = &ctxs_[thread_idx]->thread_mu_;
+    std::condition_variable* thread_cv = &ctxs_[thread_idx]->thread_cv_;
+    std::lock_guard<std::mutex> thread_lock(*thread_mu);
+    ctxs_[thread_idx]->rpc_done_ = true;
+    thread_cv->notify_one();
+  }
+
+  void IssueCallbackRpc(size_t thread_idx, Thread* t) {
+    if (!InitThreadFuncImpl(thread_idx)) {
+      return;
+    }
+
+    if (!closed_loop_) {
+      gpr_timespec next_issue_time = NextIssueTime(thread_idx);
+      // Start an alarm call back to run the internal callback after
+      // next_issue_time
+      delete ctxs_[thread_idx]->alarm_;
+      ctxs_[thread_idx]->alarm_ = new Alarm;
+      Alarm* alarm = ctxs_[thread_idx]->alarm_;
+      alarm->experimental().Set(
+          next_issue_time,
+          [this, t, thread_idx](bool ok) { ThreadFuncImpl(t, thread_idx); });
+    } else {
+      ThreadFuncImpl(t, thread_idx);
+    }
+  }
+
+ private:
+  int NumThreads(const ClientConfig& config) {
+    int num_threads = config.async_client_threads();
+    if (num_threads <= 0) {  // Use dynamic sizing
+      num_threads = cores_;
+      gpr_log(GPR_INFO, "Sizing async client to %d threads", num_threads);
+    }
+    return num_threads;
+  }
+
+  void ThreadFunc(size_t thread_idx, Thread* t) override {
+    ctxs_[thread_idx] = new CallbackClientRpcData;
+    IssueCallbackRpc(thread_idx, t);
+    WaitForCompletionOfAllRpcs(thread_idx);
+
+    NotifyMainThread(thread_idx);
+    delete ctxs_[thread_idx];
+    return;
+  }
+
+  void NotifyMainThread(size_t thread_idx) {
+    std::lock_guard<std::mutex> l(mu_);
+    threads_done_++;
+    if (threads_done_ == num_threads_) {
+      cv_.notify_one();
+    }
+  }
+
+  void WaitForCompletionOfAllRpcs(size_t thread_idx) {
+    std::mutex* thread_mu = &ctxs_[thread_idx]->thread_mu_;
+    std::condition_variable* thread_cv = &ctxs_[thread_idx]->thread_cv_;
+    std::unique_lock<std::mutex> thread_lock(*thread_mu);
+
+    while (!ctxs_[thread_idx]->rpc_done_) {
+      thread_cv->wait(thread_lock);
+    }
+    ctxs_[thread_idx]->rpc_done_ = false;
+    thread_lock.unlock();
+  }
+
+  void DestroyMultithreading() final {
+    std::unique_lock<std::mutex> l(mu_);
+    while (threads_done_ != num_threads_) {
+      cv_.wait(l);
+    }
+    EndThreads();
+  }
+};
+
+class CallbackUnaryClient final : public CallbackClient {
+ public:
+  CallbackUnaryClient(const ClientConfig& config) : CallbackClient(config) {
+    StartThreads(num_threads_);
+  }
+  ~CallbackUnaryClient() {}
+
+  bool ThreadFuncImpl(Thread* t, size_t thread_idx) override {
+    auto* stub = channels_[thread_idx % channels_.size()].get_stub();
+    ClientContext* context = new grpc::ClientContext();
+    SimpleResponse* response = new SimpleResponse();
```

Comment:
Can these be entries in the bigger `ctx` data structure so that you don't need to do separate `new` and `delete` on them?